sway-core/src/asm_generation/fuel/optimizations.rs

```diff
@@ -0,0 +1,257 @@
+use std::collections::BTreeSet;
+
+use either::Either;
+use rustc_hash::{FxHashMap, FxHashSet};
+
+use crate::{
+    asm_generation::fuel::compiler_constants,
+    asm_lang::{ControlFlowOp, VirtualImmediate12, VirtualOp, VirtualRegister},
+};
+
+use super::{
+    abstract_instruction_set::AbstractInstructionSet, analyses::liveness_analysis,
+    data_section::DataSection,
+};
+
+impl AbstractInstructionSet {
+    // Aggregates that are const index accessed from a base address
+    // can use the IMM field of LW/SW if the value fits in 12 bits.
+    pub(crate) fn const_indexing_aggregates_function(&mut self, data_section: &DataSection) {
+        // Poor man's SSA (local ... per block).
+        #[derive(PartialEq, Eq, Hash, Clone)]
+        struct VRegDef {
+            reg: VirtualRegister,
+            ver: u32,
+        }
+
+        // What does a register contain?
+        enum RegContents {
+            Constant(u64),
+            BaseOffset(VRegDef, u64),
+        }
+
+        // What is the latest version of a vreg definition.
+        let mut latest_version = FxHashMap::<VirtualRegister, u32>::default();
+        // Track register contents as we progress instructions in a block.
+        let mut reg_contents = FxHashMap::<VirtualRegister, RegContents>::default();
+
+        // Record that we saw a new definition of `reg`.
+        fn record_new_def(
+            latest_version: &mut FxHashMap<VirtualRegister, u32>,
+            reg: &VirtualRegister,
+        ) {
+            latest_version
+                .entry(reg.clone())
+                .and_modify(|ver| *ver += 1)
+                .or_insert(1);
+        }
+
+        // What's the latest definition we've seen of `reg`?
+        fn get_def_version(
+            latest_version: &FxHashMap<VirtualRegister, u32>,
+            reg: &VirtualRegister,
+        ) -> u32 {
+            latest_version.get(reg).cloned().unwrap_or(0)
+        }
+
+        for op in &mut self.ops {
+            match &mut op.opcode {
+                either::Either::Left(op) => match op {
+                    VirtualOp::ADD(dest, opd1, opd2) => {
+                        // We don't look for the first operand being a constant and the second
+                        // one a base register. Such patterns must be canonicalised prior.
+                        let Some(RegContents::Constant(c2)) = reg_contents.get(opd2) else {
+                            reg_contents.remove(dest);
+                            record_new_def(&mut latest_version, dest);
+                            continue;
+                        };
+                        match reg_contents.get(opd1) {
```

Comment:
I am thinking of the code duplication between `ADD` and `ADDI`. Perhaps eliminate it by introducing closure for the common part (`match ...`) that accepts `opd1` and `c2` as a parameters?